### PR TITLE
dataclasses.Field.default and .default_factory can be _MISSING_TYPE

### PR DESCRIPTION
**Pull request by altendky**

### CODE REVIEW

**Review by srittau — CHANGES_REQUESTED** [review]
Thanks, but please use the new union syntax (i.e. `Foo | Bar`).

File: stdlib/dataclasses.pyi [hunk 1]
```diff
@@ -71,8 +71,8 @@ class _DefaultFactory(Protocol[_T_co]):
 class Field(Generic[_T]):
     name: str
     type: Type[_T]
-    default: _T
-    default_factory: _DefaultFactory[_T]
+    default: Union[_T, _MISSING_TYPE]
```
Comment (JelleZijlstra) [inline stdlib/dataclasses.pyi:74]:
```suggestion
    default: _T | _MISSING_TYPE
```
We prefer PEP 604 syntax in new stubs.

**Comment by hauntsaninja** [comment]
I considered doing this earlier, but didn't because I think this will be slightly annoying, since the idiomatic way to narrow is `if x is dataclasses.MISSING`. mypy primer has an example, you can see that this re-introduces a type error in pydantic (that was previously fixed by https://github.com/python/typeshed/pull/5718)

**Comment by JelleZijlstra** [comment]
Yes, I had the same concern as Shantanu. I think PEP 661 (https://www.python.org/dev/peps/pep-0661/) would provide a cleaner solution.

**Comment by srittau** [comment]
Maybe @altendky has a use case that requires this change? We can always change over to PEP 661 when that gets accepted and is available for use. That said, I have no opinion, what solution is best.

**Comment by Akuli** [comment]
We could define `_MISSING_TYPE` as an enum containing a single member, named `_MISSING_TYPE.MISSING` for example, and then use `Literal[_MISSING_TYPE.MISSING]` instead of `_MISSING_TYPE`. Mypy seems to support `is` for enums, and unlike most other things, enums can go inside a `Literal`.

**Comment by altendky** [comment]
Welp, of course the time I don't chat about a change first it ends up deserving discussion.  `:]`  Here's the code that made me take a look at this.

https://github.com/python-desert/desert/blob/483734f89528fef86c0f4bbd38b4ba3473acae9d/src/desert/_make.py#L324
```python
        if field.default_factory != dataclasses.MISSING:  # type: ignore[misc]
```
```
error: Non-overlapping equality check (left operand type: "Callable[[], object]", right operand type: "_MISSING_TYPE")  [comparison-overlap]
```

Here's a minimal example.  Note that the first error seems to be https://github.com/python/mypy/issues/10750 and I was already avoiding it with `# type: ignore[misc]`.

https://mypy-play.net/?mypy=latest&python=3.10&flags=strict&gist=f66bb887158648e99b97c042165dd684
```python
import dataclasses


@dataclasses.dataclass
class C:
    f: int = 0


for field in dataclasses.fields(C):
    if field.default_factory != dataclasses.MISSING:
        pass
```
```
main.py:10: error: Attribute function "default_factory" with type "Callable[[], _T]" does not accept self argument
main.py:10: error: Non-overlapping equality check (left operand type: "Callable[[], Any]", right operand type: "_MISSING_TYPE")
Found 2 errors in 1 file (checked 1 source file)
```

I just tried `is` and `is not` and they both seemed to result in the same error.

Regardless, I'll be adding an ignore "for now" since of course this wouldn't be released anyways.  If there's another more proper solution and this isn't a useful interrim change, I can handle that.

Thanks to all for the interest and effort here.

**Review by srittau — APPROVED** [review]
This looks fine to me, but I'll defer judgement to @hauntsaninja and @JelleZijlstra.

**Comment by JelleZijlstra** [comment]
I'd prefer to stick with the current stub until PEP 661 gives us better tools to deal with this situation.

**Pull request closed without merging**